### PR TITLE
Give back our GIF category animation support since we now use plugin coders

### PR DESCRIPTION
**Pull request by dreampiggy**
Give back our GIF category animation support since we now use plugin coders. 
This will not affect FLAnimatedImageView which just create the first frame.

1. Change default coders to [IOCoder, WebPCoder]
2. Add back our previous GIF decoding code, with loop count support
3. Add GIF encoding code for animation
4. Modify IOCoder for GIF format because FLAnimatedImageView need just use the first frame

### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

Welcome back our previous `UIImage+GIF` category for animated GIF!
Because we have do refactor and separate the UIImage+XXX category with different coders. And then SD inner will not relay on that `UIImage+GIF` any methods.

This means we can keep our `FLAnimatedImageView` fast but also provide the full version `UIImage+GIF` support. We can let **ImageIOCoder** to just grab the first frame and make `FLAnimatedImageView` behave the same as previous version. And then our `UIImage+GIF` category can use **GIFCoder** to do additional GIF animation support.

Besides just grab the previous version GIF decoding code(From 3.8.2), I add some more such as image loop count(Since now we use a `sd_imageLoopCount` property). And also, I provide the animated GIF encoding method. Our users can now use that category to perform full-functional GIF support.

### CODE REVIEW

**Comment by bpoplauschi** [comment]
@dreampiggy good idea overall. 
Should we rename `SDWebImageGIFCoder` to something like `SDWebImageAnimatedGIFCoder` or `SDWebImageGIFCoreGraphicsCoder`?
Or `SDWebImageIOCoder` to `SDWebImageIOStaticCoder` ? or it just creates more confusion?
I would like to make it clear to the users which is which, but maybe it's ok how it is now.

File: SDWebImage/SDWebImageImageIOCoder.m [hunk 1]
```diff
@@ -67,7 +67,6 @@ + (instancetype)sharedCoder {
 - (BOOL)canDecodeFromData:(nullable NSData *)data {
     switch ([NSData sd_imageFormatForImageData:data]) {
         // Do not support GIF and WebP decoding
```
Comment (bpoplauschi) [inline SDWebImage/SDWebImageImageIOCoder.m]: should update this comment

File: SDWebImage/SDWebImageImageIOCoder.m [hunk 2]
```diff
@@ -373,7 +383,6 @@ - (nullable UIImage *)sd_decompressedAndScaledDownImageWithImage:(nullable UIIma
 - (BOOL)canEncodeToFormat:(SDImageFormat)format {
     switch (format) {
         // Do not support GIF and WebP encoding
```
Comment (bpoplauschi) [inline SDWebImage/SDWebImageImageIOCoder.m]: should update this comment

**Comment by bpoplauschi** [comment]
Should also update this comment from `SDWebImageIOCoder.h`:
```Built in coder that supports PNG, JPEG, TIFF, includes support for progressive decoding``` to include one-frame GIF

File: SDWebImage/SDWebImageImageIOCoder.m (hunk 2 above)
Comment (bpoplauschi) [reply SDWebImage/SDWebImageImageIOCoder.m]: Done via a32ab28

File: SDWebImage/SDWebImageImageIOCoder.m (hunk 1 above)
Comment (bpoplauschi) [reply SDWebImage/SDWebImageImageIOCoder.m]: Done via a32ab28

**Comment by bpoplauschi** [comment]
I did update the comments via a32ab28 so I can merge this PR.

**Comment by dreampiggy** [comment]
@bpoplauschi I'm a little late. But thanks for your review with this comments. 😃

**Comment by dreampiggy** [comment]
@bpoplauschi I have a idea. That this GIF image encoding processing can let it work for creating Animated Images for whatever images on macOS. Such as **Animated WebP**. I found that [](https://github.com/chuganzy/HCImage-BPG/blob/master/HCImage%2BBPG/HCImage%2BBPG.mm#L157) have already use this idea. (What I previously not even take awareness n :))

Therefore, I think I should try to achieve this. I think I should make it more universal and abstract because there is no difference if you have a array of `imageRef` with its `frameDuration`. I can make two new Helper function to decoding animated images and encoding animated images for all platform without copy and paste code anywhere. Just refactor code. The interface may looks:


```objective-c
/**
 Return animated image with images array which applyed the durtaion for each frame fix and used for UIAnimatedImage for UIKit or NSImage for AppKit
 The reason we need this fix is that UIAnimatedImage does not support setting duration for each frame
 It just use the average duration by dividing the total duration, so we need this fix to keep animated images work

 @param images The original images array
 @param durations The durations array for each image, the count must be equal to images array
 @param totalDuration The total duration for animated image, in millisecond
 @return The animated image
 */
FOUNDATION_EXPORT UIImage * SDAnimatedImageCreateWithImagesAndDurations(NSArray<UIImage *>* images, int const * const durations, NSTimeInterval totalDuration);

/**
 Return the original images array which un-applyed the duration for each frame fix and output durations
 This is the reversed method for `SDAnimatedImageCreateWithImagesAndDurations`

 @param animatedImage The applyed fix animated UIImage with images array or animated NSImage
 @param totalDuration The total duration for animated image, in millisecond
 @param durations The pointer to the durations array output
 @return The un-applyed images array
 */
FOUNDATION_EXPORT NSArray<UIImage *>* SDCreateImagesAndDurationsFromAnimatedImage(UIImage * animatedImage, NSTimeInterval totalDuration, int * const durations);
```

Or do you have a better idea ? Maybe create an new class called `SDImageFrame` which have two properties `image(or imageRef)`, `duration` is better ?


```objective-c

@interface SDImageFrame

@property (nonatomic, strong) CGImageRef imageRef; // current frame's imageRef
@property (nonatomic, assign) int duration; // current frame's duration, in milliseconds
@property (nonatomic, strong) CFDictionaryRef property; // current frame's extra information, such as thumbnail, meta information

@end

// then we can change the interface to this ...
FOUNDATION_EXPORT UIImage * SDAnimatedImageCreateWithFrames(NSArray<SDImageFrame *> *frames, int totalDuration);
FOUNDATION_EXPORT NSArray<SDImageFrame *>* SDCreateFramesFromAnimatedImage(UIImage * animatedImage, int *totalDuration);
```

**Pull request merged**